### PR TITLE
[WIP] Add mock golden rollback test with DB

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

Relates to #1567, #1575 

# Overview

- [x] Use a mock chain to easily test corner cases
- [x] Demonstrate that `totalBalance` may appear to "duplicate" money due to inclusion of change from invalid pending tx
- [ ] TODO: Factor out and create more tests

# Comments

- Abstracting further to some ultra-concise DSL might be useful
- Seems alarming to me that we don't have any properties that prevented the `totalBalance` problem. Want to understand our divergence from the spec better. But this is an orthogonal issue to this PR.

### CODE REVIEW

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 1]
```diff
@@ -447,10 +465,204 @@ fileModeSpec =  do
...
+                getAvailableBalance db `shouldReturn` 2
+                getTotalBalance db `shouldReturn` 2
+                getTxsInLedger db `shouldReturn` [(Outgoing, 3), (Incoming, 5)]
+                getPending db `shouldReturn` []
+
+
+
+                atomically $ void $ unsafeRunExceptT $ rollbackTo testPk (SlotId 0 0)
+                getTxsInLedger db `shouldReturn` []
+                -- getPending `shouldReturn` []
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:530]: I disagree here. getPending should return the outgoing transaction made at slot 2

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 2]
```diff
@@ -447,10 +465,204 @@ fileModeSpec =  do
...
+                -- but got: [(Outgoing, 3)]
+
+                getAvailableBalance db `shouldReturn` 0
+                -- getTotalBalance `shouldReturn` 0
+                -- but got: 2
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:535]: `2` is also expected here.

File: lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs [hunk 3]
```diff
@@ -447,10 +465,204 @@ fileModeSpec =  do
...
+                                (blockHash "tx2b")
+                                [ (TxIn (blockHash "tx1") 0, Coin 5)
+                                ]
+                                [ TxOut (ourAddrs !! 3) (Coin 3)
+                                , TxOut (ourAddrs !! 2) (Coin 2)
+                                ]
+                            ]
+                -- THESE ARE NOT EXPECTED, BUT PASS!!
+                -- getAvailableBalance `shouldReturn` 5
+                -- getTotalBalance `shouldReturn` 7
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:554]: I don't quite see why they "are not expected" though. This is 100% expected, but I agree, maybe not the best behavior we would like here. But, the code is simply doing doing what's it told to do.

**Comment by Anviking** [comment]
Closing, but I think we need to re-visit the wallet's behaviour regarding pending txs, balance and rollback.

_[thread with some notes / thoughts](https://input-output-rnd.slack.com/archives/GBT05825V/p1587052833056000?thread_ts=1586967724.044500&cid=GBT05825V)_
